generic-ec/src/multiscalar/straus.rs

```diff
@@ -0,0 +1,327 @@
+    ///
+    /// Preallocates memory to fit `capacity` amount of scalars
+    fn new(w: usize, capacity: usize) -> Self {
+        assert!((2..=8).contains(&w));
+        let naf_size = Scalar::<E>::serialized_len() * 8 + 1;
+        let width = 1 << w;
+
+        Self {
+            naf_size,
+            w,
+            width,
+            width_half: 1 << (w - 1),
+            matrix: Vec::with_capacity(naf_size * capacity),
+            window_mask: width - 1,
+            _curve: Default::default(),
+        }
+    }
+    /// Adds a scalar into matrix
+    fn add_scalar(&mut self, scalar: &Scalar<E>) {
+        let scalar_bytes = scalar.to_le_bytes();
+        let mut x_u64 = vec![0u64; scalar_bytes.len() / 8 + 1];
+        read_le_u64_into(&scalar_bytes, &mut x_u64[0..4]);
+
+        let offset = self.matrix.len();
+        debug_assert!(
+            offset + self.naf_size <= self.matrix.capacity(),
+            "unnecessary allocations detected"
+        );
+        self.matrix.resize(offset + self.naf_size, 0i8);
+        let naf = &mut self.matrix[offset..];
+
+        let mut pos = 0;
+        let mut carry = false;
+        while pos < self.naf_size {
+            let u64_idx = pos / 64;
+            let bit_idx = pos % 64;
+            let bit_buf: u64 = if bit_idx < 64 - self.w {
+                // This window bits are contained in a single u64
+                (x_u64[u64_idx] >> bit_idx) & self.window_mask
+            } else {
+                // Combine the current u64's bits with the bits from the next u64
+                ((x_u64[u64_idx] >> bit_idx) | (x_u64[u64_idx + 1] << (64 - bit_idx)))
+                    & self.window_mask
+            };
+
+            // Add the carry into the current window
+            let window = if carry { bit_buf + 1 } else { bit_buf };
+
+            if window & 1 == 0 {
+                // If the window value is even, preserve the carry and continue.
+                // Why is the carry preserved?
+                // If carry == 0 and window & 1 == 0, then the next carry should be 0
+                // If carry == 1 and window & 1 == 0, then bit_buf & 1 == 1 so the next carry should be 1
+                pos += 1;
+                continue;
+            }
+
+            if window < self.width_half {
+                carry = false;
+                naf[pos] = window as i8;
+            } else {
+                carry = true;
+                naf[pos] = (window as i8).wrapping_sub(self.width as i8);
+            }
+
+            pos += self.w;
+        }
+
+        debug_assert!(!carry);
+    }
+
+    /// Iterates over scalars NAF representations in the same order as
+    /// scalars were added into the matrix
+    fn iter(&self) -> impl Iterator<Item = &[i8]> {
+        self.matrix.chunks_exact(self.naf_size)
+    }
+}
+
+/// Read one or more u64s stored as little endian bytes.
+///
+/// ## Panics
+/// Panics if `src.len() != 8 * dst.len()`.
+fn read_le_u64_into(src: &[u8], dst: &mut [u64]) {
```

Comment:
Suggestion: read either le or be bytes depending on platform. I also tried to write a transmute with unsafe casts, but since alignment is not guaranteed, in the worst case we would have to copy the whole slice anyway.

Reply:
I would prefer not writing unsafe code when possible. Does this optimization has noticeable effect on performance?

Reply:
So do I correctly understand that you're suggesting to do `scalar.to_le_bytes()` or `scalar.to_be_bytes()` depending on the platform and then do `transmute_aligned_native_u64`? Doesn't that reorder u64 on different platforms? If `u64`s are in different order on different platforms, that would additionally complicate the NAF conversion function.

Reply:
Also, did you notice that we always have a trailing zero u64 in the end of vec of u64? Which means we would anyway need to allocate memory to have this trailing `0_u64`

Reply:
It's probably possible to rewrite the algorithm so it doesn't require trailing zero, but it's again an additional complication to the algorithm

Reply:
> Also, did you notice that we always have a trailing zero u64 in the end of vec of u64 

I did after I wrote my option =) Yeah, that might eat away the gains from transmuting.

What I suggest at the minumum is to have two functions: `read_le_u64_into` and `read_be_u64_into`, and we serialize the scalar to correct byte order and dispatch between the two based on the platform

Reply:
Yes? Because converting from be to be is a noop, and converting from be to le requires shuffling bytes around

Reply:
I don't really like this idea because this complicates the algorithm (now we need to serialize scalars in be/le depending on the platform, then rearrange `u64`s, or change the algorithm so we accept native order, but output little-endian), and then we can't test it (we probably can set up CI with both little-endian and big-endian instances, but that's additional work to do)

Reply:
I don't think the profit is worth it

Reply:
Ok, I agree for testability